fix #9098 dashboard filter capabilities

MV88 opened the pull request:
## Description
<!-- A few sentences describing the overall goals of the pull request' s commits. -->

Added the possibility to add spatial filter during creation of certain of widgets (charts, table, counter)

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue

**What is the current behavior?**
<!-- You can also link to an existing issue here -->
fix #9098 dashboard filter capabilities

**What is the new behavior?**
<!-- Describe here the new behaviour based on your changes -->

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information

offtherailz commented on web/client/utils/WidgetsUtils.js:
```diff
@@ -331,3 +331,73 @@ export const getSelectedWidgetData = (widget = {}) => {
+    limits: {},
+    mousePointer: 'pointer',
+    mapStateSource: 'mapWizard',
+    resolution: 4891.96981025128,
+    resolutions: [
```
are these necessary? please remove everything that is not necessary.

offtherailz commented on web/client/plugins/queryPanelWithMap/SpatialFilter.jsx:
```diff
@@ -0,0 +1,52 @@
+        spatialField: state.queryform.spatialField,
+        showDetailsPanel: state.queryform.showDetailsPanel,
+        spatialPanelExpanded: state.queryform.spatialPanelExpanded,
+        zoom: (getMapConfigSelector(state) || {}).zoom,
+        projection: (getMapConfigSelector(state) || {}).projection
```
The difference between this component and the original one stays only in these 2 selectors connection.
I think we should try to avoid the duplication introducing 2 plugins, but only 1 with some config/auto-config options.

MV88 replied:
what do you mean exactly?

offtherailz replied:
In general I think you can work on a single plugin. 
I didn't think how to switch between using the embedded map or the map plugin. 
One solution can be via a configuration flag. Anyway if it is able to find out if the map plugin is present, it could be a smarter solution.

offtherailz commented on web/client/plugins/queryPanelWithMap/SmartQueryForm.jsx:
```diff
@@ -0,0 +1,148 @@
+            resetCrossLayerFilter,
+            toggleMenu: (rowId, status) => toggleMenu(rowId, status,  "crossLayer")
+        }, dispatch),
+        controlActions: bindActionCreators({onToggleQuery: toggleControl.bind(null, 'queryPanelWithMap', null)}, dispatch)
```
Also in this case the difference is only this controlAction (that may be unified) and maybe the diffferent imports.

offtherailz replied:
I mean we can use the `queryPanel` control in any case, when we have one plugin only

offtherailz commented on web/client/components/data/query/QueryBuilder.jsx:
```diff
@@ -216,7 +216,7 @@ class QueryBuilder extends React.Component {
                 {this.renderItems('end', { spatialOperations, spatialMethodOptions, ...toolsOptions })}
             </BorderLayout>
-            : <div style={{margin: "0 auto", width: "60px"}}><Spinner spinnerName="three-bounce" overrideSpinnerClassName="spinner"/></div>;
+            : <div className="spinner-panel"style={{margin: "0 auto", width: "60px"}}><Spinner spinnerName="three-bounce" overrideSpinnerClassName="spinner"/></div>;
```
```suggestion
            : <div className="spinner-panel" style={{margin: "0 auto", width: "60px"}}><Spinner spinnerName="three-bounce" overrideSpinnerClassName="spinner"/></div>;
```

offtherailz commented on web/client/components/data/query/SpatialFilter.jsx:
```diff
@@ -216,8 +216,8 @@ class SpatialFilter extends React.Component {
                 {this.props.spatialField.method
-                     && this.getMethodFromId(this.props.spatialField.method)
-                     && this.getMethodFromId(this.props.spatialField.method).type === "wfsGeocoder"
+        && this.getMethodFromId(this.props.spatialField.method)
+        && this.getMethodFromId(this.props.spatialField.method).type === "wfsGeocoder"
```
```suggestion
                     && this.getMethodFromId(this.props.spatialField.method)
                     && this.getMethodFromId(this.props.spatialField.method).type === "wfsGeocoder"
```

offtherailz commented on web/client/epics/dashboard.js:
```diff
@@ -85,12 +85,12 @@ export const handleDashboardWidgetsFilterPanel = (action$, {getState = () => {}}
             featureTypeSelected(...getFTSelectedArgs(getState())),
             loadFilter(getEditingWidgetFilter(getState())),
-            setControlProperty('queryPanel', "enabled", true)
+            setControlProperty('queryPanelWithMap', "enabled", true)
```
I think you can keep the originals

offtherailz commented on web/client/epics/dashboard.js:
```diff
@@ -85,12 +85,12 @@ export const handleDashboardWidgetsFilterPanel = (action$, {getState = () => {}}
             Rx.Observable.race(
                 action$.ofType(QUERY_FORM_SEARCH).take(1),
-                action$.ofType(TOGGLE_CONTROL).filter(({control, property} = {}) => control === "queryPanel" && (!property || property === "enabled")).take(1)
+                action$.ofType(TOGGLE_CONTROL).filter(({control, property} = {}) => control === "queryPanelWithMap" && (!property || property === "enabled")).take(1)
```
I think you can keep the originals

offtherailz commented on web/client/epics/dashboard.js:
```diff
@@ -108,7 +108,7 @@ export const handleDashboardWidgetsFilterPanel = (action$, {getState = () => {}}
             .concat(
                 Rx.Observable.of(// drawSupportReset(),
-                    setControlProperty('queryPanel', "enabled", false)
+                    setControlProperty('queryPanelWithMap', "enabled", false)
```
I think you can keep the originals

MV88 replied:
this would mean to use one single plugin

offtherailz commented on web/client/epics/widgets.js:
```diff
@@ -301,7 +312,37 @@ export const onWidgetCreationFromMap = (action$, store) =>
+    action$.ofType(EDITOR_CHANGE)
+        .filter(({key}) => key === 'chart-layers' && isDashboardEditing(store.getState()))
+        .switchMap(() => {
+            let observable$ = Rx.Observable.empty();
```
This "default" value is never used.

MV88 replied:
ok I will reformat, but note that this codestyle was copied from the precedent epic

offtherailz reviewed (changes requested):
There are too few differences between the QueryPanel plugin and QueryPanelWithMap, not enough to create 2 distinct plugins.
Keep them separated will required too much maintainance and testing. So please try to properly unify them. 
If you need help with it, we can check. The condition of using or not using the map could be to check the presence of the map plugin or via configuration. 

Note: I didn't tested it fully because I'd like to spend this time when the full plugin is ready.

MV88 commented:
> There are too few differences between the QueryPanel plugin and QueryPanelWithMap, not enough to create 2 distinct plugins. Keep them separated will required too much maintainance and testing. 

ok, but you asked me to separate them from the start and we also discussed that creating it's own plugin for the dashboard with a different behaviour was fine (initially). now I will separate them even if the reason of the splitting was due to the different connect function and the map component included.

> So please try to properly unify them. If you need help with it, we can check. The condition of using or not using the map could be to check the presence of the map plugin or via configuration.

Imo the condition of showing it should not be the Map plugin presence, but it should be driven by the object map being populated or not, which what is present now.

> 
> Note: I didn't tested it fully because I'd like to spend this time when the full plugin is ready.
> 
> Anyway my first test was failing. In particuar The map is blank ( tested with http://localhost:8081/#/dashboard/21694 connected to dev env )
> 
>  screencast-localhost_8081-2023.10.09-10_47_12.webm 
> So please always double check:

will check

> * You committed everything, and do a final test before to ask for review

will check

> * You didn't committed anything in excess (e.g. wrong / unnecessary indentation changes of some files not involved and so on.

I mean you can close an eye if I reorder a bit some of the imported packages. this is not going to create conflicts and if it so it is something very naive to resovle.

offtherailz reviewed (changes requested):
I reorganized a little bit the components. 
- I moved the `useEmbeddedMap` to properties to `toolsOptions` and documented in JSDOC
- Now we have another target in querypanel called "map". The map is rendered there through the `SpatialFilterMap` component, if `toolsOptions.useEmbeddedMap=== true`.  I also documented this propertly
- The `SpatialFilter` component will render one or the other component, with the proper connections, depending on the `toolsOptions.useEmbeddedMap`, removing all the duplications that may confuse and be hard to maintain.
- [ ] viewport option do not work/ is not updated. 

Everything works as before, it looks, anyway I see some errors in the tests of querypanel I didn't expect after this change. I'd like to take a look at that test failure later.

- [ ] The issue I noticed (also before my changes) is that if you try to edit the spatial filter (clicking on the pencil), the modification to bbox coordinates or radius/center of circle are not updated to the map. 
- [ ] See also the comments inline.

MV88 commented on web/client/plugins/querypanel/SpatialFilterMap.jsx:
```diff
@@ -0,0 +1,44 @@
+ */
+export const MapComponent = connect(
+    createSelector([
+        getWidgetLayer,
```
hi, @offtherailz 
i see your point, but the only way to isolate the updates to the queryform state would be to:
- create new queryform actions that update queryform reducer  (creating maps and layers)
- intercept within an epic widgets edit action or layer selection action nad use previous created actions to sync the query form state

alternatively we can move the widgets actionCreator type inside queryform.

or do you have other ideas?

offtherailz replied:
the action that opens the query form can contain an optional map configuration, that uses the defaults + the layer to use.
This map will be loaded in the embedded map state. When the query panel closes, nothing else is required, only the filter have to be updated.

I think you don't have to intercept anything else.

MV88 commented on web/client/themes/default/less/dashboard.less:
```diff
@@ -35,6 +35,31 @@
         position: relative;
         overflow: auto;
     }
+    .query-form-root {
+        width: 100% !important;
+        .query-form-panel-container {
+            .spinner-panel{
+                width: 600px !important;
+            }
+            width: 100% !important;
+            transition: unset !important;
+            .mapstore-query-builder {
+                position: unset;
+                display: flex;
+                #query-form-panel {
+                    width: 600px !important;
+                }
+                .mapstore-query-map {
+                    #__base_map__ {
+                        margin: 0 !important;
+                        height: 100% !important;
+                    }
+                    flex: 1
+                }
+
+            }
+        }
+    }
```
i know we should not used them but these were needed to override certain inline styles

offtherailz replied:
solved

MV88 commented:
@offtherailz please review again

>  viewport option do not work/ is not updated.

removed this option

> The issue I noticed (also before my changes) is that if you try to edit the spatial filter (clicking on the pencil), the modification to bbox coordinates or radius/center of circle are not updated to the map.

fixed with calculation fo correct zoom in  the epic

> it was not redrawing the filter

fixed with **onMapReady** that triggers initQueryPanel

MV88 commented:
ah i have added also this part to the migration guidelines

![image](https://github.com/geosolutions-it/MapStore2/assets/11991428/9ac55cba-c3e4-4726-b1a4-bec174d1b538)

offtherailz commented on docs/developer-guide/mapstore-migration-guide.md:
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
 
+### Adding spatial filter to dashboard widgets
+
+In order to enable the possibility to add in and the spatial filter to the widgets [#9098](https://github.com/geosolutions-it/MapStore2/issues/9098) you have to edit the QueryPanel config in `localConfig.json` file adding:
```
```suggestion
In order to enable the possibility to add in and the spatial filter to the widgets ( see [#9098](https://github.com/geosolutions-it/MapStore2/issues/9098) ) you have to edit the QueryPanel config in `localConfig.json` file adding:
```

offtherailz commented on docs/developer-guide/mapstore-migration-guide.md:
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
+            "useEmbeddedMap": true
+        },
+        "spatialPanelExpanded": false,
+        "spatialOperations": [
+            {"id": "INTERSECTS", "name": "queryform.spatialfilter.operations.intersects"},
+            {"id": "CONTAINS", "name": "queryform.spatialfilter.operations.contains"},
+            {"id": "WITHIN", "name": "queryform.spatialfilter.operations.within"}
+        ],
+        "spatialMethodOptions": [
+            {"id": "BBOX", "name": "queryform.spatialfilter.methods.box"},
+            {"id": "Circle", "name": "queryform.spatialfilter.methods.circle"},
+            {"id": "Polygon", "name": "queryform.spatialfilter.methods.poly"}
+        ],
```
Is it possible to avoid one of these? I know that here viewport will be removed, but at least we can use the defaults for spatialOperations?

MV88 replied:
this would mean to alter the current default props? do you want this?
without the spatial operations you get this now
![image](https://github.com/geosolutions-it/MapStore2/assets/11991428/929c4df6-10d3-4136-a15c-dc644c269eef)

offtherailz replied:
Ah, ok. So the defaults are very old and probably wrong, and they should be updatd. Spatial Operations too?

offtherailz replied:
I tested it personally. Without specifying it, the tool do not work. This is a limitation of the query panel, but not part of this PR. 
Maybe we will make the query panel have a better default configuration, without needing to specify

offtherailz commented on web/client/plugins/QueryPanel.jsx:
```diff
@@ -263,7 +263,7 @@ class QueryPanel extends React.Component {
 
     UNSAFE_componentWillReceiveProps(newProps) {
-        if (newProps.queryPanelEnabled === true && this.props.queryPanelEnabled === false) {
+        if (!newProps.toolsOptions.useEmbeddedMap && newProps.queryPanelEnabled === true && this.props.queryPanelEnabled === false) {
```
Why this change ?

MV88 replied:
because I wanted to avoid to double init_query_panel

offtherailz commented on docs/developer-guide/mapstore-migration-guide.md:
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
+
+In order to enable the possibility to add in and the spatial filter to the widgets [#9098](https://github.com/geosolutions-it/MapStore2/issues/9098) you have to edit the QueryPanel config in `localConfig.json` file adding:
+
+- **useEmbeddedMap** flag,
+- **spatialOperations**
+- **spatialMethodOptions**
```
```suggestion
- **useEmbeddedMap**: flag to enable the embedded map
- **spatialOperations**: The list of spatial operations allowed for this plugin
- **spatialMethodOptions**: the list of spatial methods to use.
```

offtherailz commented on docs/developer-guide/mapstore-migration-guide.md:
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
+- **spatialOperations**
+- **spatialMethodOptions**
+
+  *Note: these props can vary, for example you can limit the spatial filtering operation to intersect only*
```
```suggestion
```

offtherailz commented on docs/developer-guide/mapstore-migration-guide.md:
````diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
+
+```json
+...
+dashboard: [
````
```suggestion
"dashboard": [
```

offtherailz commented on web/client/epics/widgets.js:
```diff
@@ -322,44 +324,30 @@ export const onWidgetCreationFromMap = (action$, store) =>
+
+export const onResetMapEpic = (action$, store) =>
+    action$.ofType(TOGGLE_CONTROL)
+        .filter((type, control) => !queryPanelSelector(store.getState()) && control === "queryPanel" || isDashboardEditing(store.getState()))
```
This toggles always the map editor. You maybe mean && ? 

```suggestion
        .filter((type, control) => !queryPanelSelector(store.getState()) && control === "queryPanel" && isDashboardEditing(store.getState()))
```

MV88 commented:
@offtherailz can you approve now?

MV88 commented:
@ElenaGallo please test it in DEV

ElenaGallo commented:
Hi @MV88 I found two issues that also concern the _Attribute Filter_ option, @tdipisa it is better to open a new issue for these?

- The Attribute filter and Area of interest do not interact with each other and cannot be used at the same time. Use [this map](https://dev-mapstore.geosolutionsgroup.com/mapstore/#/dashboard/46426)
https://github.com/geosolutions-it/MapStore2/assets/56537133/b7216d85-670f-43fd-92c3-c7b097b4daf5

- The Attribute filter is not visible on the filter map
![filter map](https://github.com/geosolutions-it/MapStore2/assets/56537133/27674bad-b4bf-4f6a-8cd6-540637b65701)

tdipisa commented:
@ElenaGallo thank you. I double checked and I can confirm these two issues. @ElenaGallo please create a new dedicated issue with same labels to be assigned to @MV88.

ElenaGallo commented:
@tdipisa new issue here: #9605
